### PR TITLE
Try to use posix_fadvise with CBufferedFile

### PR DESCRIPTION
**Pull request by luke-jr**
Resurrecting #12491, since it seemed fine and desirable...

### CODE REVIEW

**Comment by MeshCollider** [comment]
Concept ACK, would be good to hear from @eklitzke why the original was closed though

File: src/util.h [hunk 1]
```diff
@@ -79,6 +79,15 @@ bool RenameOver(fs::path src, fs::path dest);
...
 bool DirIsWritable(const fs::path& directory);
 
+//! Return the original FILE* unchanged. On POSIX systems that support it,
+//! also advise the kernel that the file will be accessed sequentially.
+FILE* AdviseSequential(FILE *file);
+
+//! Close a file and return the result of fclose(). On POSIX systems that
+//! support it, advise the kernel to remove the file contents from the page
+//! cache (which can help on memory-constrained systems).
+int CloseAndDiscard(FILE *file);
```
Comment (laanwj) [inline src/util.h]: "discard" sounds to me as if writes to the file will be discarded, too
maybe `CloseAndUncache`?

**Comment by laanwj** [comment]
Concept ACK though I'd really like to see benchmarks for changes like this, to know if this has any impact in practice worth making the risk/complication of the change.

**Comment by jgarzik** [comment]
Probably worthing having -- rough concept ACK -- however:

* The double-lseek will be slower than fstat() on some networked or detachable filesystems.

* If close-and-uncache is used on just-written-to files, this action is a no-op unless a sync (fdatasync etc.) preceded it.

* Definitely worth benchmarking on multiple filesystems to see the positive or negative value of this change.

**Comment by luke-jr** [comment]
Rebased and addressed @laanwj's function rename request.

File: src/util/system.cpp [hunk 2]
```diff
@@ -1118,6 +1109,49 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
...
+        int fd = fileno(file);
+        if (fd == -1)
+            goto close;
+        end = lseek(fd, 0, SEEK_END);
+        if (end == -1)
+            goto close;
+        posix_fadvise(fd, 0, end, POSIX_FADV_DONTNEED);
+    }
+#endif
+ close:
```
Comment (practicalswift) [inline src/util/system.cpp]: Move `close:` inside `#if … #endif`?

Comment (luke-jr) [reply src/util/system.cpp]: Replaced goto with nested ifs.

**Comment by luke-jr** [comment]
Rebased

File: src/util/system.cpp [hunk 3]
```diff
@@ -17,16 +17,6 @@
...
-// for posix_fallocate
-#ifdef __linux__
-
-#ifdef _POSIX_C_SOURCE
-#undef _POSIX_C_SOURCE
-#endif
-
-#define _POSIX_C_SOURCE 200112L
-
-#endif // __linux__
```
Comment (vasild) [inline src/util/system.cpp]: I think we shouldn't remove this. We are setting `_POSIX_C_SOURCE` in order to ask the system to provide the functions. [feature_test_macros(7)](https://linux.die.net/man/7/feature_test_macros) explains that indeed we are supposed to set it ourselves (not just observe its default value). And, of course, set it before including the system headers.

Further, we should consider OSes that will not provide the functions we need (`posix_fadvise()` in this case) even if we define `_POSIX_C_SOURCE` to `200112` (e.g. MacOS). So, a [configure-time check](https://github.com/bitcoin/bitcoin/pull/18437/files#diff-67e997bcfdac55191033d57a16d1408aR854-R867) is warranted, similarly to what https://github.com/bitcoin/bitcoin/pull/18437 does.

File: src/util/system.cpp [hunk 4]
```diff
@@ -1111,6 +1102,47 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
...
+        return file;
+    off_t start = lseek(fd, 0, SEEK_CUR);
+    if (start == -1)
+        return file;
+    off_t end = lseek(fd, 0, SEEK_END);
+    if (end != -1) {
+        posix_fadvise(fd, start, end - start, POSIX_FADV_WILLNEED);
+        posix_fadvise(fd, start, end - start, POSIX_FADV_SEQUENTIAL);
+    }
+    lseek(fd, start, SEEK_SET);
```
Comment (vasild) [inline src/util/system.cpp]: We can avoid seeking to the end of the file and back just to get the file size. Supplying `0` as the 3rd argument will suffice. From [posix_fadvise(2)](https://linux.die.net/man/2/posix_fadvise):

> ... for len bytes (or until the end of the file if len is 0)

File: src/util/system.cpp [hunk 5]
```diff
@@ -1111,6 +1102,47 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
...
+}
+
+int CloseAndUncache(FILE *file) {
+#if _POSIX_C_SOURCE >= 200112L
+    // Ignore any errors up to and including the posix_fadvise call since it's
+    // advisory.
+    if (file != nullptr) {
+        const int fd = fileno(file);
+        if (fd != -1) {
+            const off_t end = lseek(fd, 0, SEEK_END);
```
Comment (vasild) [inline src/util/system.cpp]: nit: I like the `const`, but it is missing from variables in `AdviseSequential()`, maybe add it there too?

File: src/util/system.cpp [hunk 6]
```diff
@@ -1111,6 +1102,47 @@ void AllocateFileRange(FILE *file, unsigned int offset, unsigned int length) {
...
+            if (end != (off_t)-1) {
```
Comment (vasild) [inline src/util/system.cpp]: The typecast is not needed as `off_t` is "[extended signed integral](https://pubs.opengroup.org/onlinepubs/7908799/xsh/systypes.h.html)".

**Review by vasild — COMMENTED** [review]
My gut feel tells me that this is good, but experience tells me that it needs some performance testing to demonstrate that it really brings a boost. Sometimes such changes have surprisingly adverse effects or no effects and end up being just code clutter.

File: src/util/system.cpp (hunk 6 above)
Comment (luke-jr) [reply src/util/system.cpp:1137]: Prefer to stick to the manpage.

**Comment by luke-jr** [comment]
Rebased & comments addressed

**Comment by practicalswift** [comment]
Adding to the three previous requests for benchmarks:

Do we have any benchmarks showing the practical gains in swiftness? :)

**Comment by aureleoules** [comment]
This change seems to slow down IBD.
I benchmarked by syncing a fresh node both on master and this PR (rebased on master) until block height=495000 with `-connect=localhost` (syncing from existing local full node).
Master took 5h27mins and this PR took 5h40mins.

**Comment by vasild** [comment]
This is about 4% difference. Could it be within noise/variance? I mean if you run 3 times `master` and it takes e.g. 5h27min, 5h03min, 5h42min then you know the result from the PR is within noise. I think that if this makes some impact (positive or negative) it should be observable on smaller tests too, e.g. something that runs 5 or 10 minutes. Then you can repeat a few times and see how much is the variance for both `master` and the PR.

Is such a workload disk bound? If it is, then does the full node (the source of the sync) compete for disk resources too (that would slow down the syncing node and bring noise)? Was the full node also running the PR?

**Comment by aureleoules** [comment]
@vasild you're right, sorry about my last benchmark. I didn't lock the CPU frequency which probably caused a lot of noise.

I re-ran some benchmarks with CPU frequency locked. I rebased the PR locally on master.
The local node used to sync the benchmarked nodes runs on this PR.

<details>
<summary>PR: -prune=10000 -stopatheight=100000</summary>

```
Command being timed: "./src/bitcoind -datadir=/tmp/btc -prune=10000 -stopatheight=100000 -connect=localhost -server=0"
        User time (seconds): 42.94
        System time (seconds): 13.05
        Percent of CPU this job got: 113%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:49.29
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 489628
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 10
        Minor (reclaiming a frame) page faults: 157231
        Voluntary context switches: 551434
        Involuntary context switches: 732
        Swaps: 0
        File system inputs: 1512
        File system outputs: 582728
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

</details>

<details>
<summary>master: -prune=10000 -stopatheight=100000</summary>

```
Command being timed: "./bitcoin/src/bitcoind -datadir=/tmp/btc -prune=10000 -stopatheight=100000 -connect=localhost -server=0"
        User time (seconds): 43.54
        System time (seconds): 13.47
        Percent of CPU this job got: 114%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:49.74
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 489436
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 0
        Minor (reclaiming a frame) page faults: 157198
        Voluntary context switches: 559451
        Involuntary context switches: 1026
        Swaps: 0
        File system inputs: 80
        File system outputs: 582712
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

</details>

<details>
<summary>PR: -prune=10000 -stopatheight=200000</summary>

```
Command being timed: "./src/bitcoind -datadir=/tmp/btc -prune=10000 -stopatheight=200000 -connect=localhost -server=0 -printtoconsole=0"
        User time (seconds): 266.81
        System time (seconds): 38.87
        Percent of CPU this job got: 120%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:13.04
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 815832
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 1
        Minor (reclaiming a frame) page faults: 245274
        Voluntary context switches: 1974226
        Involuntary context switches: 2469
        Swaps: 0
        File system inputs: 704
        File system outputs: 8404056
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

</details>

<details>
<summary>master: -prune=10000 -stopatheight=200000</summary>

```
Command being timed: "./src/bitcoind -datadir=/tmp/btc -prune=10000 -stopatheight=200000 -connect=localhost -server=0 -printtoconsole=0"
        User time (seconds): 268.12
        System time (seconds): 41.59
        Percent of CPU this job got: 120%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 4:15.96
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 814496
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 1
        Minor (reclaiming a frame) page faults: 245239
        Voluntary context switches: 1990059
        Involuntary context switches: 2580
        Swaps: 0
        File system inputs: 624
        File system outputs: 8376800
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0
```

</details>

The results seem to be similar, so ACK 3f2c08b8202e6a98cc55ff101e274281b54c64f1.

**Comment by achow101** [comment]
This PR does not seem to have conceptual support. Please leave a comment if you would like this to be reopened.

**Pull request closed without merging**